crates/uv-tool/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+    #[error("Failed to find a directory for executables")]
+    NoExecutableDirectory,
+}
+
+/// A collection of uv-managed tools installed on the current system.
+#[derive(Debug, Clone)]
+pub struct InstalledTools {
+    /// The path to the top-level directory of the installed tools.
+    root: PathBuf,
+}
+
+impl InstalledTools {
+    /// A directory for installed tools at `root`.
+    fn from_path(root: impl Into<PathBuf>) -> Self {
+        Self { root: root.into() }
+    }
+
+    /// Prefer, in order:
+    /// 1. The specific tool directory specified by the user, i.e., `UV_TOOL_DIR`
+    /// 2. A directory in the system-appropriate user-level data directory, e.g., `~/.local/uv/tools`
+    /// 3. A directory in the local data directory, e.g., `./.uv/tools`
+    pub fn from_settings() -> Result<Self, Error> {
+        if let Some(tool_dir) = std::env::var_os("UV_TOOL_DIR") {
+            Ok(Self::from_path(tool_dir))
+        } else {
+            Ok(Self::from_path(
+                StateStore::from_settings(None)?.bucket(StateBucket::Tools),
+            ))
+        }
+    }
+
+    pub fn tools_toml_path(&self) -> PathBuf {
+        self.root.join("tools.toml")
+    }
+
+    /// Return the toml tracking installed tools.
+    pub fn toml(&self) -> Result<ToolsToml, Error> {
+        match fs_err::read_to_string(self.tools_toml_path()) {
+            Ok(contents) => Ok(ToolsToml::from_string(contents)
+                .map_err(|err| Error::TomlRead(self.tools_toml_path(), Box::new(err)))?),
+            Err(err) if err.kind() == io::ErrorKind::NotFound => Ok(ToolsToml::default()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    pub fn toml_mut(&self) -> Result<ToolsTomlMut, Error> {
+        let toml = self.toml()?;
+        ToolsTomlMut::from_toml(&toml).map_err(|err| Error::TomlEdit(self.tools_toml_path(), err))
+    }
+
+    pub fn find_tool_entry(&self, name: &str) -> Result<Option<Tool>, Error> {
+        let toml = self.toml()?;
+        Ok(toml.tools.and_then(|tools| tools.get(name).cloned()))
+    }
+
+    pub fn add_tool_entry(&self, name: &str, tool: &Tool) -> Result<(), Error> {
+        let mut toml_mut = self.toml_mut()?;
+        toml_mut
+            .add_tool(name, tool)
+            .map_err(|err| Error::TomlEdit(self.tools_toml_path(), err))?;
+
+        // Save the modified `tools.toml`.
+        fs_err::write(self.tools_toml_path(), toml_mut.to_string())?;
+
+        Ok(())
+    }
+
+    pub fn create_environment(
+        &self,
+        name: &str,
+        interpreter: Interpreter,
+    ) -> Result<PythonEnvironment, Error> {
+        let environment_path = self.root.join(name);
+
+        debug!(
+            "Creating environment for tool `{name}` at {}.",
+            environment_path.user_display()
+        );
+
+        // Discover an interpreter.
+        // Note we force preview on during `uv tool run` for now since the entire interface is in preview
+
+        // Create a virtual environment.
```

Comment:
Hm?

Reply:
Yeah this copy pasta is no good. Removed in a later commit but 'll drop here too.